src/riak_kv_sweeper_fold.erl

```diff
@@ -0,0 +1,641 @@
+    successful_sweep(Succ, Index),
+    failed_sweep(Failed, Index).
+
+-spec successful_sweep([#sweep_participant{}], riak_kv_sweeper:index()) -> [].
+successful_sweep(Succ, Index) ->
+    [Module:successful_sweep(Index, FinalAcc) ||
+       #sweep_participant{module = Module, acc = FinalAcc} <- Succ].
+
+-spec failed_sweep([#sweep_participant{}], riak_kv_sweeper:index()) -> [].
+failed_sweep(Failed, Index) ->
+    [Module:failed_sweep(Index, Reason) ||
+       #sweep_participant{module = Module, fail_reason = Reason } <- Failed].
+
+failed_sweep(Failed, Index, Reason) ->
+    [Module:failed_sweep(Index, Reason) ||
+       #sweep_participant{module = Module} <- Failed].
+
+-spec format_result(#sa{}) -> sweep_result().
+format_result(#sa{swept_keys = SweptKeys,
+                  active_p = Succ, failed_p = Failed}) ->
+    {SweptKeys,
+     format_result(succ, Succ) ++ format_result(fail, Failed)}.
+
+-spec format_result('succ'|'fail', [#sweep_participant{}]) ->
+    [{riak_kv_sweeper:participant_module(), 'succ' | 'fail'}].
+format_result(SuccFail, Results) ->
+    [{Module, SuccFail} || #sweep_participant{module = Module} <- Results].
+
+%% Throttle depending on swept keys.
+-spec maybe_throttle_sweep(binary(), #sa{}) -> #sa{}.
+maybe_throttle_sweep(_RObjBin, #sa{throttle = {pace, _, _}} = Acc0) ->
+    #sa{throttle = {pace, Limit, Wait},
+        swept_keys = SweepKeys,
+        throttle_total_wait_msecs = TotalWait} = Acc0,
+    case SweepKeys rem Limit of
+        0 ->
+            %% We use receive after to throttle instead of sleep.
+            %% This way we can respond on requests while throttling
+            Acc1 = Acc0#sa{throttle = get_sweep_throttle(),
+                           throttle_total_wait_msecs = TotalWait + Wait},
+            timer:sleep(Wait),
+            maybe_extra_throttle(Acc1);
+        _ ->
+            maybe_extra_throttle(Acc0)
+    end;
+
+%% Throttle depending on total obj_size.
+maybe_throttle_sweep(RObjBin, #sa{throttle = {obj_size, _, _}} = Acc0) ->
+    #sa{throttle = {obj_size, Limit, Wait},
+        total_obj_size = TotalObjSize,
+        throttle_total_wait_msecs = TotalWait} = Acc0,
+    ObjSize = byte_size(RObjBin),
+    TotalObjSize1 = ObjSize + TotalObjSize,
+    case (Limit =/= 0) andalso (TotalObjSize1 > Limit) of
+        true ->
+            %% We use receive after to throttle instead of sleep.
+            %% This way we can respond on requests while throttling
+            Acc1 = Acc0#sa{throttle = get_sweep_throttle(),
+                           throttle_total_wait_msecs = TotalWait + Wait},
+            timer:sleep(Wait),
+            maybe_extra_throttle(Acc1#sa{total_obj_size = 0});
+        _ ->
+            maybe_extra_throttle(Acc0#sa{total_obj_size = TotalObjSize1})
+    end.
+
+%% Throttle depending on how many objects the sweep modify.
+-spec maybe_extra_throttle(#sa{}) -> #sa{}.
+maybe_extra_throttle(Acc) ->
```

Comment:
I meant that instead of pattern matching on the first line of the function move that pattern match into the function head.